src/js/program/program-controller.js

```diff
@@ -4,34 +4,61 @@ import cancelable from 'utils/cancelable';
 import { MediaControllerListener } from 'program/program-listeners';
 import Events from 'utils/backbone.events';
 import BackgroundMedia from 'program/background-media';
-import { PLAYER_STATE, STATE_BUFFERING, STATE_IDLE, STATE_PAUSED } from 'events/events';
+import { PLAYER_STATE, STATE_IDLE, STATE_BUFFERING, STATE_PAUSED } from 'events/events';
 import { PlayerError, MSG_CANT_PLAY_VIDEO, ERROR_PLAYLIST_ITEM_MISSING_SOURCE } from 'api/errors';
+import { ItemPromise } from 'controller/next-item-promise';
+import { normalizePlaylistItem } from 'playlist/playlist';
+import Item from 'playlist/item';
 
 class ProgramController extends Events {
     /**
      * ProgramController constructor
      * @param {Model} model - The player's model
      * @param {MediaElementPool} mediaPool - The player's media element pool
+     * @param {Api} api - The player API applied as context in async playlist item promise callback execution
      */
-    constructor(model, mediaPool) {
+    constructor(model, mediaPool, api) {
         super();
 
         this.adPlaying = false;
+        this.api = api;
         this.background = BackgroundMedia();
         this.mediaPool = mediaPool;
         this.mediaController = null;
         this.mediaControllerListener = MediaControllerListener(model, this);
         this.model = model;
         this.providers = new Providers(model.getConfiguration());
-        this.loadPromise = Promise.resolve();
+        this.loadPromise = null;
         this.backgroundLoading = model.get('backgroundLoading');
+        this.itemPromises = [];
 
         if (!this.backgroundLoading) {
             // If background loading is not supported, set the shared media element
             model.set('mediaElement', this.mediaPool.getPrimedElement());
         }
     }
 
+    asyncItem(index, item) {
+        const { api, model } = this;
+        // Set the player state to buffering if there is a playlist item callback
+        const deferBufferingState = setTimeout(() => {
+            model.set(PLAYER_STATE, STATE_BUFFERING);
+        });
+        return this.getItemPromise(index, api).run().then((playlistItem) => {
+            clearTimeout(deferBufferingState);
+            if (playlistItem && playlistItem !== item) {
```

Comment:
@pajong what if this line is changed to 
`if (playlistItem && playlistItem !== this.model.get('playlist')[index]) {`?

Reply:
The issue is not the check here - the issue is the two lines following this line.
`this.setPlaylistItem(index, playlistItem)` does:
`const newItem = normalizePlaylistItem(model, new Item(item), item.feedData || {});
        if (newItem) {
            model.get('playlist')[index] = newItem;
            return newItem;
        }
}`
This returned `newItem` is but thrown away, and `playlistItem` (without the normalize) is being used to set activeItem in `media-controller`

Reply:
Thanks for catching this @pajong. I've made the following changes and added some tests to cover item replacements and background loading transition https://github.com/jwplayer/jwplayer/pull/3595/commits/8a2756844342a79fbe14cc96e26ff93576ac19d4